File: src/types/graph.ts
export interface Entity {
    id: string;
    name: string;
    type: EntityType;
    description?: string;
    metadata: Record<string, any>;
    createdAt: string;
    updatedAt: string;
}

export interface Relationship {
    id: string;
    sourceId: string;
    targetId: string;
    type: RelationshipType;
    properties?: Record<string, any>;
    confidence?: number;
    notes?: string;
    createdAt: string;
}

export type EntityType =
    | 'domain'
    | 'data_product'
    | 'process'
    | 'person'
    | 'technology'
    | 'ai_agent'
    | 'metadata_technical'
    | 'product_category'
    | 'region'
    | 'supply_chain_node'
    | 'marketing_channel'
    | 'kpi'
    | 'legal_entity'
    | 'finance_entity'
    | 'policy'
    | string;
export type RelationshipType = string;

export interface GraphData {
    entities: Entity[];
    relationships: Relationship[];
}

export interface GraphNode extends Entity {
    x?: number;
    y?: number;
    fx?: number | null;
    fy?: number | null;
}

export interface GraphLink extends Relationship {
    source: string | GraphNode;
    target: string | GraphNode;
}

export interface GraphFilter {
    entityTypes?: EntityType[];
    relationshipTypes?: RelationshipType[];
    searchQuery?: string;
    /** When set, scopes the graph to exactly this entity ID set (e.g. a chat response's used context) */
    entityIds?: string[];
}

File: backend/models/graph.py
from pydantic import BaseModel
from typing import Dict, Any, Optional
from datetime import datetime

class EntityBase(BaseModel):
    name: str
    type: str
    description: Optional[str] = None
    metadata: Dict[str, Any] = {}

class EntityCreate(EntityBase):
    pass

class EntityUpdate(BaseModel):
    name: Optional[str] = None
    type: Optional[str] = None
    description: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None

class Entity(EntityBase):
    id: str
    createdAt: str
    updatedAt: str

    model_config = {
        "from_attributes": True,
        "extra": "ignore"
    }

class RelationshipBase(BaseModel):
    sourceId: str
    targetId: str
    type: str
    properties: Optional[Dict[str, Any]] = None
    confidence: Optional[float] = None
    notes: Optional[str] = None

class RelationshipCreate(RelationshipBase):
    pass

class Relationship(RelationshipBase):
    id: str
    createdAt: str

    model_config = {
        "from_attributes": True,
        "extra": "ignore"
    }

class GraphData(BaseModel):
    entities: list[Entity]
    relationships: list[Relationship]
